### PR TITLE
checkpointing support

### PR DESCRIPTION
**Pull request by jchiang87**
@danielsf I think this is basically finished.  I tested these changes with `imsim.py` and everything seems to work, but I still would like to put some unit tests here, which I hope to do today or tomorrow.  Please have a look, and let me know if you see any problems.

### CODE REVIEW

**Comment by cwwalter** [comment]
Wow. That was fast!  I just started to look at it. Two questions:

- Since it is counting by objects how are the background photons considered?

- How does it decide whether or not to start up with a checkpoint file?  We would drive this decision from imSim?

**Comment by cwwalter** [comment]
> Regarding whether the checkpoint file is written, the GalSimInterpreter class has a .checkpoint_file attribute. If that is None, it doesn't do any checkpointing. There are options in imsim.py to set that file and the cadence for writing it.

Ah.  OK. I hadn't seen the other PR before.  Actually I was asking a slightly different question.  Let's say there is a checkpoint file lying around.  How do we decide whether to start with it or not?

**Comment by jchiang87** [comment]
If we provide imsim.py with the filename of an existing checkpoint file, it will read it in and proceed from the checkpoint.  If we provide a name for a file that does not exist, the code will start from scratch and write the checkpoint file.  In both cases, it will update (overwrite) that file at the specified cadence.

**Comment by cwwalter** [comment]
Ah, I see.  OK. Nice solution.

I imagine we might want the option to start from scratch even if the file already exists.  Do you want to just handle that outside of the program?

**Comment by jchiang87** [comment]
This is ready for review.

File: tests/testGalSimInterface.py [hunk 1]
```diff
@@ -991,6 +992,102 @@ def testPSF(self):
...
+
+    def make_detector(self, detname):
+        return self.detectors[detname]
+
+class CheckPointingTestCase(unittest.TestCase):
+    """
+    TestCase class for testing the GalSimInterpreter checkpointing functions.
+    """
+    def setUp(self):
+        self.cp_file = 'checkpoint_test.pkl'
```
[the remark this answers is not in the document]
Comment (jchiang87) [reply tests/testGalSimInterface.py]: thanks. will do.

**Pull request merged**